atuin-client/src/record/sqlite_store.rs

```diff
@@ -122,29 +123,75 @@ impl Store for SqliteStore {
     }
 
     async fn get(&self, id: RecordId) -> Result<Record<EncryptedData>> {
-        let res = sqlx::query("select * from records where id = ?1")
-            .bind(id.0.as_simple().to_string())
+        let res = sqlx::query("select * from store where store.id = ?1")
+            .bind(id.0.as_hyphenated().to_string())
             .map(Self::query_row)
             .fetch_one(&self.pool)
             .await?;
 
         Ok(res)
     }
 
-    async fn len(&self, host: HostId, tag: &str) -> Result<u64> {
-        let res: (i64,) =
-            sqlx::query_as("select count(1) from records where host = ?1 and tag = ?2")
-                .bind(host.0.as_simple().to_string())
+    async fn last(&self, host: HostId, tag: &str) -> Result<Option<Record<EncryptedData>>> {
+        let res =
+            sqlx::query("select * from store where host=?1 and tag=?2 order by idx desc limit 1")
+                .bind(host.0.as_hyphenated().to_string())
                 .bind(tag)
+                .map(Self::query_row)
                 .fetch_one(&self.pool)
+                .await;
+
+        match res {
+            Err(sqlx::Error::RowNotFound) => Ok(None),
+            Err(e) => Err(eyre!("an error occured: {}", e)),
+            Ok(record) => Ok(Some(record)),
+        }
+    }
+
+    async fn first(&self, host: HostId, tag: &str) -> Result<Option<Record<EncryptedData>>> {
+        self.idx(host, tag, 0).await
+    }
+
+    async fn len(&self, host: HostId, tag: &str) -> Result<Option<u64>> {
+        let last = self.last(host, tag).await?;
+
+        if let Some(last) = last {
+            return Ok(Some(last.idx + 1));
+        }
+
+        return Ok(None);
+    }
```

Comment:
a bit weird that empty doesn't return 0

Reply:
true! At the time I was thinking absense of store vs length 0, but should totally simplify that